plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/FileSystemExchangeSink.java

```diff
@@ -0,0 +1,241 @@
+    private volatile boolean finished;
+    @GuardedBy("this")
+    private CompletableFuture<Void> blockedFuture = new CompletableFuture<>();
+
+    public FileSystemExchangeSink(FileSystemExchangeStorage exchangeStorage, URI outputDirectory, int outputPartitionCount, Optional<SecretKey> secretKey)
+    {
+        this.exchangeStorage = requireNonNull(exchangeStorage, "exchangeStorage is null");
+        this.outputDirectory = requireNonNull(outputDirectory, "outputDirectory is null");
+        this.outputPartitionCount = outputPartitionCount;
+        this.numBuffers = outputPartitionCount * 2; // double buffering to overlap computation and I/O
+        this.secretKey = requireNonNull(secretKey, "secretKey is null");
+        this.writeBufferSize = exchangeStorage.getWriteBufferSizeInBytes();
+
+        for (int i = 0; i < numBuffers; ++i) {
+            freeBuffers.add(Slices.allocate(writeBufferSize).getOutput());
+        }
+    }
+
+    // The future returned by {@link #isBlocked()} should be considered as a best-effort hint. It's
+    // possible to return NOT_BLOCKED when it's actually blocked in certain cases. The decision is
+    // consciously made to avoid global locking.
+    @Override
+    public CompletableFuture<?> isBlocked()
+    {
+        if (freeBuffers.isEmpty() && pendingBuffers.size() < outputPartitionCount) {
+            synchronized (this) {
+                if (blockedFuture.isDone()) {
+                    blockedFuture = new CompletableFuture<>();
+                }
+                return blockedFuture;
+            }
+        }
+        else {
+            return NOT_BLOCKED;
+        }
+    }
+
+    @Override
+    public void add(int partitionId, Slice data)
+    {
+        throwIfFailed();
+
+        checkArgument(partitionId < outputPartitionCount, "partition id is expected to be less than %s: %s", outputPartitionCount, partitionId);
+        checkState(!finished, "already finished");
+        if (finished) {
+            return;
+        }
+
+        writers.computeIfAbsent(partitionId, this::createWriter);
+        synchronized (writers.get(partitionId)) {
+            writeToExchangeStorage(partitionId, Slices.wrappedIntArray(data.length()));
+            writeToExchangeStorage(partitionId, data);
+        }
+    }
+
+    private ExchangeStorageWriter createWriter(int partitionId)
+    {
+        URI outputPath = outputDirectory.resolve(partitionId + DATA_FILE_SUFFIX);
+        try {
+            return exchangeStorage.createExchangeStorageWriter(outputPath, secretKey);
+        }
+        catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    private void writeToExchangeStorage(int partitionId, Slice slice)
+    {
+        int position = 0;
+        while (position < slice.length()) {
+            SliceOutput pendingBuffer = pendingBuffers.computeIfAbsent(partitionId, ignored -> {
+                try {
+                    return freeBuffers.take();
```

Comment:
> When freeBuffers.take() is called after abort this will deadlock

All write futures have been registered with a callback to return the buffer back to `freeBuffers`, so a deadlock shouldn't be possible. Am I missing something obvious?

Reply:
Yeah signal works, but we can also use a poison to signal abortion like `EMPTY_SLICE.getSliceOutput()`. I find it easier to reason about concurrency. Let me know what you think.

Reply:
Both way works, but I feel wait-notify is better. Let me implement it.